### PR TITLE
LineInformation Request and Response service 

### PR DESCRIPTION
**Pull request by ue71603**
Specially for the shape of lines in the map service.
If more information from NeTEx should be used, please advise. The Siri is a bit limited as well. After a first discussion, I will add an example as well.

fixes: https://github.com/VDVde/OJP/issues/208

### CODE REVIEW

File: OJP/OJP_Lines.xsd [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+				<xs:annotation>
+					<xs:documentation>DIRECTION of LINE if relevant for the shape.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="Shape" type="siri:LineShapeStructure" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>The shape of the line. Normally a line has multiple routes. This is just the main one for presentation.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="Probability" minOccurs="0">
```
Comment (herlitze) [inline OJP/OJP_Lines.xsd:44]: How do you suppose to get this probability? Does this refer to the route of the line, or the line itself?

File: OJP/OJP_Lines.xsd [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+					<xs:documentation>Reference to the LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element ref="siri:PublishedLineName" minOccurs="0"/>
+			<xs:element name="DirectionRef" type="siri:DirectionRefStructure" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>DIRECTION of LINE if relevant for the shape.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="Shape" type="siri:LineShapeStructure" minOccurs="0">
```
[the remark this answers is not in the document]
Comment (ue71603) [reply OJP/OJP_Lines.xsd:39]: @Aurige : Could you answer?

File: OJP/OJP_Lines.xsd (hunk 1 above)
Comment (ue71603) [reply OJP/OJP_Lines.xsd:44]: @herlitze  Should I name this a score? You do scores also for stops. I also would think this is the hidden ingredient for trip planners.

Comment (herlitze) [reply OJP/OJP_Lines.xsd:44]: @ue71603 There is only a LineDirectionFilterStructure in the request, so the probability refers to the route returned, correct?
Is the probability something like (number of trips of this line going this route)/(number of total trips of the line), or what would you expect here?
Is LineDirectionFilterStructure the correct request structure? What is an excluded line supposed to do?

Comment (ue71603) [reply OJP/OJP_Lines.xsd:44]: * A LineDirection can have different routes. Perhaps it is only one, perhaps there are several one. If you think a trip planner can say nothing, then we remove the probability. Frankly I have only a vague idea, if this is useful. I just copied things from the PlaceResults
* LineDirectionFilterStructure: Not that line. Again: Your expert judgement is needed, to say, if we need a new structure or not. I just reused things.

File: OJP/OJP_Lines.xsd (hunk 2 above)
[the remark this answers is not in the document]
Comment (ue71603) [reply OJP/OJP_Lines.xsd:39]: Aurige: Would prefer NeTEx and use of gml. So here only GML. GML is very heavy payload, because of that. Google has a a lighter polyline code. Proposal using GML will be made.

File: OJP/OJP_Lines.xsd (hunk 1 above)
Comment (ue71603) [reply OJP/OJP_Lines.xsd:44]: Malte: Only ID then we don't need a score.
==> remove the PRobability.

**Comment by skinkie** [comment]
Still itches. Especically 'main one' and 'direction if relevant'. I guess for a high detail route it is always relevent, unless this is just an indicative presentation.

**Review by skinkie — DISMISSED** [review]
If we would have a 'next branch' this would certainly be in. Only merge if we really want to have this feature.

**Comment by ue71603** [comment]
Stephan: Why GML was used. We have a different system for showing the projection. 
Matthias: check LegTrackStructure and do it the same way.

File: OJP/OJP_Lines.xsd [hunk 3]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<xs:schema xmlns="http://www.vdv.de/ojp" xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:siri="http://www.siri.org.uk/siri" xmlns:gml="http://www.opengis.net/gml/3.2" targetNamespace="http://www.vdv.de/ojp" elementFormDefault="qualified" attributeFormDefault="unqualified" id="OJP_siri_service">
+	<xs:annotation>
+		<xs:documentation>OJP/OJP_Lines.xsd - OJP Line information request and response</xs:documentation>
+	</xs:annotation>
+	<xs:import namespace="http://www.siri.org.uk/siri" schemaLocation="../siri/siri_model/siri_all-v2.0.xsd"/>
+	<xs:include schemaLocation="OJP_All.xsd"/>
+	<xs:import namespace="http://www.siri.org.uk/siri" schemaLocation="../siri/siri_model/siri_modes-v1.1.xsd"/>
```
Comment (sgrossberndt) [inline OJP/OJP_Lines.xsd:8]: Only one siri import should be necessary

Comment (ue71603) [reply OJP/OJP_Lines.xsd]: done

File: OJP/OJP_Lines.xsd [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+					<xs:documentation>Reference to the LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element ref="siri:PublishedLineName" minOccurs="0"/>
+			<xs:element name="DirectionRef" type="siri:DirectionRefStructure" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>DIRECTION of LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="LinkProjection" type="LinkProjectionStructure" minOccurs="0">
```
Comment (Aurige) [inline OJP/OJP_Lines.xsd:38]: Project is expected to be on something in TM (usually on the infrastructure roard/rail)
The concept of "main geometry" sounds blurry to me and does not always exist.
Is it expected to be oriented (event if there is no direction)?
"an each" should be "and each"

Comment (ue71603) [reply OJP/OJP_Lines.xsd:38]: @Aurige : What should I correct? You disagree with Projection? We have this already in the Legs. The typo will be corrected.

Comment (Aurige) [reply OJP/OJP_Lines.xsd:38]: A Projection should refer to what it projects on ... If you would refer to some OMS (or equivalent) features, that would be Ok.
But, if I understand well, what is described here is the Route geometry (or part of it) ... So RouteGeometry would be better for the naming.

And on the functional side, I'm stille puzzled by the possible definition of "main geometry": what would you use for the RER A in Paris ? https://www.paris-metro-map.info/paris-rer-a-map/

Comment (ue71603) [reply OJP/OJP_Lines.xsd]: @Aurige How would you model this in Transmodel or NeTEx? I guess it is not possible either as this is done per Journey. Lines like that should better not exist in my view. If we want to show it that way, then I would need to model this with a completly new structure which probably needs to be a full graph (because circularity is possible as well). 
A possibility would be to allow multiple routes (then RER A is modeled as 6 routes.) What do you think?

Comment (ue71603) [reply OJP/OJP_Lines.xsd]: @Aurige Or do you wish me to expand the service that it also provides the line as a representation of stops (and then really doing the projection)?

Comment (Aurige) [reply OJP/OJP_Lines.xsd]: > A possibility would be to allow multiple routes (then RER A is modeled as 6 routes.) What do you think?

yes that's the way to go (I agree that one should avoid such geometry... but it does exist ans is comunicated to the public... no way to change this now)

Comment (Aurige) [reply OJP/OJP_Lines.xsd]: In fact you want a simple OJP service for a structure that can be complex in NeTEx ... at some point you just need to refer the NeTEx data.

File: OJP/OJP_Lines.xsd [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+					<xs:documentation>Reference to the LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element ref="siri:PublishedLineName" minOccurs="0"/>
+			<xs:element name="DirectionRef" type="siri:DirectionRefStructure" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>DIRECTION of LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="RouteGeometry" type="LinkProjectionStructure" minOccurs="0">
```
Comment (Aurige) [inline OJP/OJP_Lines.xsd:38]: Name the type **RouteGeometryStructure** (or **LinearShapeStructure** if you want to be generic), and make it unbounded, and that's Ok for me (with a comment "see NeTEx data for the detailed structure")

File: OJP/OJP_Lines.xsd (hunk 4 above)
Comment (ue71603) [reply OJP/OJP_Lines.xsd]: ok. opted for the reduced version.

File: OJP/OJP_Lines.xsd (hunk 5 above)
Comment (ue71603) [reply OJP/OJP_Lines.xsd]: switched to LinearShapeStructure and unbounded

File: OJP/OJP_Requests.xsd [hunk 6]
```diff
@@ -203,7 +203,6 @@
 			<xs:extension base="AbstractOJPServiceRequestStructure">
 				<xs:sequence>
 					<xs:group ref="TripRefineRequestGroup"/>
-					<xs:element ref="siri:Extensions" minOccurs="0"/>
```
Comment (skinkie) [inline OJP/OJP_Requests.xsd:206]: This is an unrelated change. Must not be in this pull request.

File: OJP/OJP_Requests.xsd [hunk 7]
```diff
@@ -247,4 +246,31 @@
...
+			</xs:extension>
+		</xs:complexContent>
+	</xs:complexType>
+	<xs:element name="OJPLineInformationDelivery" type="OJPLineInformationDeliveryStructure" substitutionGroup="siri:AbstractFunctionalServiceDelivery"/>
+	<xs:complexType name="OJPLineInformationDeliveryStructure">
+		<xs:complexContent>
+			<xs:extension base="siri:AbstractServiceDeliveryStructure">
+				<xs:sequence>
+					<xs:element ref="OJPLineInformationRequest" minOccurs="0"/>
+					<!-- <xs:group ref="ServiceResponseContextGroup"/> no Context for this service so far-->
```
Comment (skinkie) [inline OJP/OJP_Requests.xsd:269]: Mmm.... I don't know if a comment is better than leaving it out.

**Review by skinkie — CHANGES_REQUESTED** [review]
As above.

File: OJP/OJP_Requests.xsd (hunk 7 above)
Comment (ue71603) [reply OJP/OJP_Requests.xsd:270]: @skinkie  your choice. I remove it, if you want me to.

File: OJP/OJP_Requests.xsd (hunk 6 above)
Comment (ue71603) [reply OJP/OJP_Requests.xsd]: ok will remove it.

Comment (ue71603) [reply OJP/OJP_Requests.xsd]: ... from this PR...

File: OJP/OJP_Requests.xsd [hunk 8]
```diff
@@ -247,4 +247,31 @@
...
```
Comment (ue71603) [inline OJP/OJP_Requests.xsd:270]:
```suggestion
```

File: OJP/OJP_Common.xsd [hunk 9]
```diff
@@ -47,14 +47,14 @@
...
+	<xs:complexType name="LinearShapeStructure">
 		<xs:annotation>
-			<xs:documentation>An oriented correspondence from one LINK of a source layer, onto an entity in a target layer: e.g. LINK SEQUENCE.</xs:documentation>
+			<xs:documentation>An oriented correspondence from one LINK or ROUTE of a source layer, onto an entity in a target layer: e.g. LINK SEQUENCE. As OJP is reduced in relation to NeTEx, we use a simple general structure.</xs:documentation>
 		</xs:annotation>
 		<xs:sequence>
 			<xs:element name="Position" type="siri:LocationStructure" minOccurs="2" maxOccurs="unbounded">
 				<xs:annotation>
-					<xs:documentation>Ordered list of locations representing the geogemtry of the link.</xs:documentation>
+					<xs:documentation>Ordered list of locations representing the geogemtry of the link/route.</xs:documentation>
```
Comment (skinkie) [inline OJP/OJP_Common.xsd:57]:
```suggestion
					<xs:documentation>Ordered list of locations representing the geometry of the link or route.</xs:documentation>
```

**Review by skinkie — APPROVED** [review]

**Review by Aurige — DISMISSED** [review]

File: OJP/OJP_Lines.xsd [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+			</xs:element>
+			<xs:element ref="siri:PublishedLineName" minOccurs="0"/>
+			<xs:element name="DirectionRef" type="siri:DirectionRefStructure" minOccurs="0">
+				<xs:annotation>
+					<xs:documentation>DIRECTION of LINE.</xs:documentation>
+				</xs:annotation>
+			</xs:element>
+			<xs:element name="RouteGeometry" type="LinearShapeStructure" minOccurs="0" maxOccurs="unbounded">
+				<xs:annotation>
+					<xs:documentation>The LINE's route geometry. A line can have multiple ROUTEs and each has its own geometry. The first should be the "main" geometry..</xs:documentation>
```
Comment (skinkie) [inline OJP/OJP_Lines.xsd:40]:
```suggestion
					<xs:documentation>The LINE's route geometry. A line can have multiple ROUTEs and each has its own geometry. The first should be the "main" geometry.</xs:documentation>
```

**Review by skinkie — APPROVED** [review]

**Review by AndreasAtSBB — APPROVED** [review]

**Review by AndreasAtSBB — APPROVED** [review]
OJP_Lines.xsd added
Switched from "LinkProjectionStructure" to "LinkaShapeStructure"

**Comment by sgrossberndt** [comment]
@skinkie Please delete merged branches

**Comment by sgrossberndt** [comment]
@skinkie Please also shorten the commit messages when doing squash and merge. All those interim commit messages are not that helpful IMO: https://github.com/VDVde/OJP/commit/0f6cda8f8852dafb3f1eddd9051dba25e8444ac6

**Comment by skinkie** [comment]
@sgrossberndt will do.

**Review by trurlurl — COMMENTED** [review]
In my view LineProblem is missing as part of LineResultStructure. Or are there good reasons not to have it?

**Comment by ue71603** [comment]
I will do this in an additional PR (added an issue): https://github.com/VDVde/OJP/issues/300
The PR was older than the adding of problem. This can't be helped.
